runtime/onert/core/src/ir/train/TrainableGraph.cc

```diff
@@ -233,6 +230,46 @@ std::vector<ir::OperationIndex> TrainableGraph::btopolSortOperations() const
   return ret;
 }
 
+std::vector<ir::OperationIndex>
+TrainableGraph::truncateBackwardOrder(std::vector<ir::OperationIndex> backward_order) const
+{
+  auto forward_order = backward_order;
+  std::reverse(forward_order.begin(), forward_order.end());
+
+  std::set<ir::OperationIndex> alive;
+
+  for (const auto &index : forward_order)
+  {
+    const auto &op = operations().at(index);
+    const auto &trainable_op = dynamic_cast<const ITrainableOperation &>(op);
+
+    if (trainable_op.hasTrainableParameter())
+      alive.insert(index);
+
+    // TODO: replace this with `std::set::contains` after C++20
+    if (alive.find(index) != alive.end())
+    {
+      for (const auto &output : op.getOutputs())
+      {
+        const auto &operand = operands().at(output);
+        for (const auto &use : operand.getUses())
+        {
+          alive.insert(use);
+        }
+      }
+    }
+  }
+
+  backward_order.erase(std::remove_if(backward_order.begin(), backward_order.end(),
+                                      [&](const auto &index) {
+                                        // TODO: replace this with `std::set::contains` after C++20
+                                        return alive.find(index) == alive.end();
+                                      }),
+                       backward_order.end());
+
+  return backward_order;
+}
```

Comment:
I have a question. 

Does this algorithm handle the below case? 

The sequence of `[FC, has trainable parameter] -> [MaxPool, NO trainable param] -> [Reshape, No trainable param]` ?  
```
  /*
            (input)
                      ╲
            (weight) ⎼[FC]⎼>(out1)-> [MaxPool] ->(out2)-> [Reshape] -> (y_pred)
                    ╱                                                                                               ╲
               (bias)                                                                                               [Loss]⎼> (output)
                                                                                                                      ╱
                                                                                                          (y_true)
  */
```

Reply:
Aha, I re-read the code and it's test cases.

Since `No trainable param Node` also is inserted in `Alive` set, `Alive` is propagated through the forward sequence.
 
I got it!